### ISSUE TITLE
make preparelocal use S3 for tarball and unify with --dryrun

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6544]
currently preparelocal creates a tarball with all needed stuff for executing the job wrapper
(inside DagmanCreator) called InputFiles.tar.gz and sends it to the schedd, from where crab preparelocal
fetches it to create local directory where to run the job.

Such tarball should be transferred via S3 cache instead, and possibly with same code as for --dryrun.

Even better, --dryrun should be executed inside the directory created by preparelocal, and should not be part of the submit command.

something like:
* submit --dry  : does all things like submit, uploads all tarballs, but does not submit to schedd
* preparelocal : like now, creates local dir for  a task submitted or submitted with --dry
* crab testrun (new command) runs in the local director for 10 events (like current submit --dry)

Also, currently in the schedd there are both `InputFiles.tar.gz` and `input_files.tar.gz` ! pretty damn confusing

Difficulty is to have a way to implement a piece at a time, w/o breaking things.

**Comment by belforte** [comment]
should be enough to go in this order:
1. make crab preparelocal use things from S3 and never talk to sched
2. implement crab testrun as something to do after crab preparelocal
3. change submit --dry to be stop after uploading and point user to crab testrun

**Comment by belforte** [comment]
this relates with https://github.com/dmwm/CRABServer/issues/7461

**Comment by novicecpp** [comment]
The goals of this issue as we agreed in the meeting are:
- Deprecate `crab submit --dryrun` behavior. 
    - Current behavior is we upload "custom" `InputFiles.tar.gz` to S3 and set tasks status to `UPLOADED` (see [DryRunUploader](https://github.com/dmwm/CRABServer/blob/master/src/python/TaskWorker/Actions/DryRunUploader.py)). Then, client pull this custom `inputFiles.tar.gz`, extract, and run locally. Each time it run will increase number of event it consumed and measure a time to give user idea how to adjust splitting parameter.
    - Users more likely want to check if crabConfig.py/PSet.py able to run in grid than see how long does it take *(citation needed)a*.
- Nothing change to preparelocal. Download the submitted file and prepare env for run locally. Note that to do `preparelocal`, user need to submit the real task with limited total job to run.

**Comment by novicecpp** [comment]
I would like to explain how we move input files (wrapper, scripts, and user sandbox) in crab systems first.

- First, user submit the task, client upload the sandbox to S3 (`s3://crabcache/<cernusername>/sandboxes/<sha1>.tar.gz`).
- Second, TW create `InputFiles.tar.gz` as zipped executable to transfer to schedd later https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanCreator.py#L726-L729.
    - This include user `sandbox.tar.gz` where we download back from S3 then add to the list of InputFiles https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanCreator.py#L1239-L1240 List of files in `InputFiles.tar.gz` is described in [here](https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanCreator.py#L1239-L1240)
- Third, the `InputFiles.tar.gz` will be ["submitted"](https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanSubmitter.py#L399 ) to schedd along with `subdag.jdl` (I do not know what it is) through `transfer_input_files` https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanSubmitter.py#L540 of condor submit command.
- Forth, in schedd, `dag_bootstrap_startup.sh` extract `InputFiles.tar.gz` to `SPOOL_DIR` https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/scripts/dag_bootstrap_startup.sh#L54
- Finally, when Dagman submit the job, it use `transfer_input_files` again but only upload a necessary files. You can see the list [here](https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanCreator.py#L143) and here https://github.com/dmwm/CRABServer/blob/9b4679d14bb19ccc7373d56c20631eea34d80a69/src/python/TaskWorker/Actions/DagmanCreator.py#L537-L552

**Comment by novicecpp** [comment]
So, to unify `preparelocal` and `submit --dryrun`, 
I will make `submit --dryrun` to have the similar UX as `preparelocal` but without submitting to the schedd. This functionality is for users to test their new tasks before real crab submission (Thanks to Dario for the idea and pointout the pain of `preaparelocal`).

This is why I said earlier that I want to deprecate the behavior, not remove the command.

This is possible and easy to do in client side. Thanks to Dario again for `crab recover` command that use other command like `crab status/kill` behind the scence.

Basically, `crab submit --dryrun` will submit the task to let TW create the `InputFiles.tar.gz`, wait until task status change to `UPLOADED` and then continue to trigger `preparelocal` command.

For the server side, well..a lot of code changes needed, obviously the `InputFiles.tar.gz` need to upload to S3, and spliting sandbox from `InputFiles.tar.gz` to not reupload the sandbox again and again, especially the compaign-like task where it share sandbox.

I Will write the detail down tomorrow.

**Comment by novicecpp** [comment]
And of course thanks to Stefano for the original ideas on how to unify both commands and the attempted of improving the `--dryrun` code (both client and server side) that I can reuse it to fix this issue.

**Comment by belforte** [comment]
sounds like you know more than me about this matter now :-)

`subdag.jdl` is used for automatic splitting`, we do not users want to run automatic splitting machinery interactively via `--dryrun`

**Comment by novicecpp** [comment]
Here is the pointer to the code:
Server: PR https://github.com/dmwm/CRABServer/pull/8645
Client: PR https://github.com/dmwm/CRABClient/pull/5329

These PR changes 3 things:
- Separate `sandbox.tar.gz` from `InputFiles.tar.gz` 
    - `InputFiles.tar.gz` is the file contains everything we need to run in schedd and work node.
    - This is needed to unify the code between `submit`, `submit --dryrun`, and `preparelocal`.
- `crab submit --dryrun` do `preparelocal`, but the task is not submit to schedd.
- `preparelocal --jobid 1` simply call `preparelocal` and execute in one go 
    - When I read the preparelocal code, I feel like it does not make any sense to have a different way to run the job. I might be wrong though.

Because I separated sandbox from `InputFiles.tar.gz`, I am not sure if some other commands or code have the assumption that it expects sandbox in there, beside the code I changed. 
So, it needs more test. But the simple test is passed though https://gitlab.cern.ch/crab3/CRABServer/-/pipelines/7990590 ( the branch base on [master 7f4e9eed](https://github.com/dmwm/CRABServer/commit/7f4e9eed41081d9407a0c7b838a9438e0ba5bac0)).

**Comment by novicecpp** [comment]
@belforte I have moved this task to Todo in "CRAB Work Planner". 
Feel free to bump priority as you see fit.

**Comment by belforte** [comment]
I looked at the code and have only some minor questions.
Time to lay out a plan. As a general strategy I'd rather break submit --dryrun for a while then mix old and new code. Cleaning up has always been difficult.

1. add support for ` checkS3Object` in RESTCache . This can go in immediately and deploy new REST
2. add checkS3Object  to ServerUtilities 
3. Modify AdjustSites and dag_bootstrap_startup to download sandbox from S3. Can go in at any time
4. Modify DagmanCreator to stop downloading sandbox.tar.gz and upload slimmed InputFiles.tar.gz. Can go in at any time after 3.
5. could changes to Client go in here ? Need to test
6. Modify Handler to use new DryRun.py instead of DryRunUploader.py
7. Deploy new Client if not done at 5.

**Comment by belforte** [comment]
OK. Let's deploy the change to DagmanCreator ASAP so we can push new client in production.

**Comment by belforte** [comment]
back to this. Code from Wa is in https://github.com/novicecpp/CRABServer/tree/get_sandbox_in_schedd
I have now imported that branch in my repo: https://github.com/belforte/CRABServer/tree/get_sandbox_in_schedd

**Comment by belforte** [comment]
1. and 2. done in https://github.com/dmwm/CRABServer/pull/8740

**Comment by belforte** [comment]
doing 4. I found a bug in #8740 which was only affecting sandbox existance check in new code.
Will fix in same PR as to push 4.

But I have also found that we also  still need `debug_files.tar.gz` to be  moved around via TW, which creates confusion. Should change AdjustSites.py to download that and create that debug_files directory at bootstrap time.
Better yet find a way to upload them as files to S3, like down for twlog, but that may require extensve changes to Client, ServerUtilities and RestCache. Current code supports uploading of a single tarball using objectype=debugfiles. But we want 3 separate files for  config, pset and scriptExe

**Comment by belforte** [comment]
let's go with a:

- [ ] 4.5 expand debug_files.tar.gz in AdjustSites and stop handling it in DagmanCreator

**Comment by belforte** [comment]
I decided to go in steps.
* deploy a Client compatible with new handling of tarballs via S3 https://github.com/dmwm/CRABClient/issues/5340
* deploy TW changes done until now
* do further changes for dryrun and do not worry if it stays broken until client is updated. Or maybe simply deprecate the "run" part of `submit --dryrun` and tell uses to use preparelocal and `run_job`

**Comment by belforte** [comment]
time to look at `--dryrun` .
Problem is how to make a smooth transition. Let's start with a compatibility matrix

client                | TW 241017 (prod) | TW 2410108 (preprod) |
 --                      |  --                              |  --                                     |
240930 (prod) |                                  |                                         |
241030 (last)   |                                  |                                         |

**Comment by belforte** [comment]
so currently `--dryrun` is broken since 10 days  (interesting that no user reported it).
And we way want to deploy new TW anyhow.
So I only need to fix thing for "a new client and TW 241018 (or later)"

**Comment by belforte** [comment]
New client requires to update REST.

**Comment by belforte** [comment]
REST updated :-)

**Comment by belforte** [comment]
I tried various things, but too much gets broken. Better break dryrun for `crab-pre` as well, i.e. deploy new TW and then a modified client which "resurrects it".  In the meanwhile user will have to do with `crab preparelocal`.

Main problem is that current DryRunUploader action creates as `runtimefiles` object a tarfaile which contains `InputFiles.tar.gz` and this confused the code in `prepareLocal` which expects to get directly `InputFiles.tar.gz` from S3.

Best solutions seems to me to modify `DagmanCreator` so that `splitting-summary.json` is added in `InputFiles.tar.gz` and then "nullify" `DryRunUploader` as per [this](https://github.com/dmwm/CRABServer/pull/8645/files#diff-fe07df38725c9cffd98cc5dcacbdb99710d0233651aea0a44f4f83268c4b0855) and replace with [DryRun](https://github.com/dmwm/CRABServer/pull/8645/files#diff-20bc3a8ea6990e51d0cc631c40881fd6a0252aecfc45665d988fc160e65707dd)

Better be "radical" :
  - deploy new client
  - fix TW with all new code and let dryrun stay broken until

**Comment by belforte** [comment]
I believe I converged on implementation.
We need to deploy new client v3.241030 in order to update TW.

Then I am going for:
- add an Uploader action after DagmanCreator, always. So tasks always go through the UPLOADED status
- if --dryrun, stop there, otherwise add also DagmanSubmitter action

DagmanCreator creates InputFiles.tar.gz and Uploader uploads it to S3

DagmanCreator also creates splitter-summary.json used for --dryrun and adds it to InputFilers.tar.gz

submit --dryrun is modified to use the new JSON way to pass arguments to CMSRunAnalysis.py and gets simplere and more readable, and keeps working.

my current branch https://github.com/belforte/CRABServer/tree/make-dryrun-work-again  should be all that's needed TW-side

I will further simplify by removing the --jobid option for crab preparelocal. Users will have these possibilities:
 - crab submit --dryrun : same look as now, simply runs for 25 seconds, limited usefulness
 - crab preparelocal  AND `sh run_job.sh` to do a proper test/debug of their jobs
 Everything will run in projdir/local and files will stay after command completes, no tmp directories !

**Comment by belforte** [comment]
so maybe I do not need **"my new"** TW.

But it is not acceptable that new client will fail to do `preparelocal` on tasks submitted with current TW.

Let's see if I can enhance https://github.com/belforte/CRABClient/tree/new-preparelocal-dryrun to solve [nt]

**Comment by belforte** [comment]
[nt] failure is due to running `preparelocal` before task has bootstrapped on the scheduler.
With new TW all needed files will be available in S3 after DagmanCreator has run, but with previous TW one has to wait for files to be in the WEB_DIR. As simple as "I was testing w/o waiting long enough" since now after a task is in SUBMITTED it takes > 1min for bootstrap to run due to https://github.com/dmwm/CRABServer/commit/83b8daca8fe4d35020453b704666a75605915881

**Comment by belforte** [comment]
conclusion:  https://github.com/dmwm/CRABServer/pull/8767 will close this from TW side. 
Leaving to future work cleaning up of compatibility and further reductions/simplification of all that passing around of files.

**Comment by belforte** [comment]
And of course there's no point in making `crab --dryrun submit` work with TW's deployed before that client version ! So once we deploy TW as of #8767 we can then safely update CRAB Client and everything will work

**Comment by belforte** [comment]
only remaining issue that new client with new TW prints an annoying message for `crab preparelocal` about the failure to retrieve webdirproxy information from REST until the task bootstrap has completed. Even if that info is not needed.
Same for `submit --dryrun`

**Comment by belforte** [comment]
All OK now !

Last bit of cosmetics could be to try "new" way first in `crab preparelocal` to avoid annoying message
```Failed to download 'InputFiles.tar.gz' from WEB_DIR. Will try to get tarballs from S3```
which now will be printed on all new tasks until a new client is deployed.

Well can simply remove the msg, but all in all try new way first is better and will make it more clear which code to remove later on.

**Comment by belforte** [comment]
test new client after changing to use "new way first"

client                    | TW 241017 (prod   ) | TW 241018 (preprod) | [PR 8767](https://github.com/dmwm/CRABServer/pull/8767) |
 --                          |  --                                 |  --                                     | --- | 
[my new branch](https://github.com/belforte/CRABClient/tree/new-preparelocal-dryrun) submit   |    OK    | OK                 |   OK   |   
[my new branch](https://github.com/belforte/CRABClient/tree/new-preparelocal-dryrun)    preparelocal|     FAIL [nt]      |       OK          |  OK        |               
[my new branch](https://github.com/belforte/CRABClient/tree/new-preparelocal-dryrun)  dryrun|    FAIL [eof]            |     FAIL  [eof]                  | OK  |

**Comment by belforte** [comment]
and of course the status for TW v3.241018 is only for information during this development, we are not going to deploy that, we'll go straight for a new tag with the fix from #8767

**Comment by belforte** [comment]
closed via #8767 and https://github.com/dmwm/CRABClient/pull/5347
